src/proxy-worker.ts

```diff
@@ -0,0 +1,285 @@
+								);
+								throw new errors.HandledTunnelingError(err.message);
+							})
+							.then(vpnHost => {
+								if (vpnHost.id === this.serviceId) {
+									throw new errors.HandledTunnelingError(
+										'device is not available on registered service instance',
+									);
+								}
+								this.logger.info(
+									`forwarding tunnel request for ${uuid}:${port} via ${vpnHost}`,
+								);
+								return this.forwardRequest(
+									vpnHost.ip_address,
+									uuid,
+									port,
+									auth,
+								).catch(errors.RemoteTunnellingError, err => {
+									this.logger.crit(
+										`error forwarding request for ${uuid}:${port} (${err.message})`,
+									);
+									throw new errors.HandledTunnelingError(err.message);
+								});
+							});
+					}
+				}),
+		);
+
+	public start = (port: number) => {
+		this.listen(port, () => {
+			this.logger.notice(`tunnel listening on port ${port}`);
+		});
+	};
+
+	private parseRequest = (req: nodeTunnel.Request) => {
+		if (req.url == null) {
+			throw new errors.BadRequestError();
+		}
+
+		const match = req.url.match(
+			/^([a-fA-F0-9]+)\.(balena|resin|vpn)(?::([0-9]+))?$/,
+		);
+		if (match == null) {
+			throw new errors.InvalidHostnameError(`invalid hostname: ${req.url}`);
+		}
+		const [, uuid, tld, port = '80'] = match;
+		if (tld === 'resin') {
+			this.logger.warning(`'.resin' tld is deprecated, use '.balena'`);
+		}
+
+		let auth;
+		if (req.auth != null && req.auth.password != null) {
+			auth = Buffer.from(req.auth.password);
+		}
+
+		return { uuid, port: parseInt(port, 10), auth };
+	};
+
+	private tunnelToDevice: nodeTunnel.Middleware = (
+		req,
+		cltSocket,
+		_head,
+		next,
+	) =>
+		Bluebird.try(() => {
+			const { uuid, port, auth } = this.parseRequest(req);
+			this.logger.info(`tunnel requested to device ${uuid} on port ${port}`);
+
+			// we need to use VPN_SERVICE_API_KEY here as this could be an unauthenticated request
+			return device
+				.getDeviceByUUID(uuid, VPN_SERVICE_API_KEY)
+				.tap(data => {
+					if (data == null) {
+						cltSocket.end('HTTP/1.0 404 Not Found\r\n\r\n');
+						throw new errors.HandledTunnelingError(`device not found: ${uuid}`);
+					}
+				})
+				.tap(data =>
+					device
+						.canAccessDevice(data, port, auth)
+						.tap(isAllowed => {
+							if (!isAllowed) {
+								cltSocket.end(
+									'HTTP/1.0 407 Proxy Authorization Required\r\n\r\n',
+								);
+								throw new errors.HandledTunnelingError(
+									`device not accessible: ${uuid}`,
+								);
+							}
+						})
+						.tap(() => {
+							if (!data.is_connected_to_vpn) {
+								cltSocket.end('HTTP/1.0 503 Service Unavailable\r\n\r\n');
+								throw new errors.HandledTunnelingError(
+									`device not available: ${uuid}`,
+								);
+							}
+						}),
+				)
+				.tap(() => (req.url = `${uuid}.vpn:${port}`));
+		})
+			.then(() => next())
+			.catch(errors.APIError, err => {
```

Comment:
Assuming `tunnelToDevice` is called regularly then there will be a definite cost to this sugar, so it's probably better to do:
```suggestion
			.catch(err => {
				if (err instanceof errors.APIError) {
					this.logger.alert(`Invalid Response from API (${err.message})`);
					cltSocket.end('HTTP/1.0 500 Internal Server Error\r\n\r\n');
				} else if (...) {
					...
				}
			})
```

Reply:
Really? That sucks :(

Reply:
It's a shame but at the same time the new format is how you'd have to do it with await/async so it at least makes that later transition easier